Add documentation for JLD2 I/O (#3365)
JuliaData/DataFrames.jl: docs/src/man/importing_and_exporting.md

# Importing and Exporting Data (I/O)

## JLD2 Files

A convenient format for reading and writing DataFrames in JLD2 which saves and loads
Julia data structures in a format comprising a subset of HDF5. Unlike several other 
formats, JLD2 preserves custom Types. The save and load functions, provided by FileIO, 
provide a mechanism to read/write `DataFrame` from/to a JLD2 file.  

A `DataFrame` can be saved as a JLD2 file at path `output.jld2` using
```julia

> using FileIO

Reviewer: I would assume it also requres JLD2.jl package loaded - right?

Author: JLD2 now added

# Importing and Exporting Data (I/O)

## JLD2 Files

> A convenient format for reading and writing DataFrames in JLD2 which saves and loads

Reviewer: ```suggestion
A convenient format for reading and writing data frames in JLD2, which saves and loads
```

Julia data structures in a format comprising a subset of HDF5. Unlike several other 
formats, JLD2 preserves custom Types. The save and load functions, provided by FileIO, 
provide a mechanism to read/write `DataFrame` from/to a JLD2 file.  

> A `DataFrame` can be saved as a JLD2 file at path `output.jld2` using

Reviewer: ```suggestion
A data frame can be saved as a JLD2 file output.jld2 using

```

format is not easy. Therefore CSV.jl is the preferred package to write CSV files
for data stored in data frames.

## JLD2 Files

> A convenient format for reading and writing data frames in JLD2, which saves and loads
> Julia data structures in a format comprising a subset of HDF5. Unlike several other 
> formats, JLD2 preserves custom Types. The `save` and `load` functions, provided by FileIO.jl, 
> allow to read/write a data frame from/to a JLD2 file.  

Reviewer: ```suggestion
JLD2 is a HDF5-compatible file format that allows reading and writing data frames.
A valuable feature of JLD2 format is that it preserves custom column types of the stored data frame.

The `save` and `load` functions, provided by FileIO.jl, allow to read/write a data frame from/to a JLD2 file.  
```

Author: This looks good... just an FYI FileIO.jl *or* JLD2.jl can be used for read/write of JLD2 files.. but I think it's best to point the usere to a single package FileIO.jl

Author: updated

A data frame can be saved as a JLD2 file output.jld2 using

```julia

> using FileIO

Reviewer: maybe separate this out and explain:
1. that this is a one-time action.
2. what is needed first time and what is needed later.

(especially that you assume that FileIO.jl is already installed while you install JLD2.jl)

Author: good point.. now updated

using Pkg; Pkg.add("JLD2")

df = DataFrame(x=1, y=2)
save("output.jld2", Dict("df" => df))
```

and can be read using
```julia

> load("output.jld2") # -> Dict{String, Any} with 1 entry: "df" => 1×2 DataFrame

Reviewer: it is better to comment on the returned value. This is a manual so we do not need to used comments to explain things.

format is not easy. Therefore CSV.jl is the preferred package to write CSV files
for data stored in data frames.

## JLD2 Files

> JLD2 is a HDF5-compatible file format that allows reading and writing data frames.

Reviewer 2: ```suggestion
[JLD2](https://github.com/JuliaIO/JLD2.jl) is a HDF5-compatible file format that allows
reading and writing data frames (and any Julia object).
```

> A valuable feature of JLD2 format is that it preserves custom column types of the stored data frame.

Reviewer 2: ```suggestion
A valuable feature of the JLD2 format is that it preserves custom column types of the stored data frame.
```

> The `save` and `load` functions, provided by FileIO.jl, allow to read/write a data frame from/to a JLD2 file.  

Reviewer 2: ```suggestion
The `save` and `load` functions, provided by [FileIO.jl](https://github.com/JuliaIO/FileIO.jl/),
provides convenience functions to read/write objects from/to a JLD2 file.
```

> A data frame can be saved as a JLD2 file output.jld2 using
> 

Reviewer 2: ```suggestion
```

> If you have not used the FileIO and JLD2 packages before then you may need to install it first:

Reviewer 2: ```suggestion
If you have not used the FileIO and JLD2 packages before then you may need to install them first:
```

```julia

> using Pkg; 

Reviewer 2: ```suggestion
using Pkg
```

> Pkg.add("FileIO")
> Pkg.add("JLD2")

Reviewer 2: This will typically be faster:
```suggestion
Pkg.add(["FileIO", "JLD2"])
```

```

The FileIO and JLD2 functions are not loaded automatically and must be imported into the session.
```julia

> using FileIO
> using JLD2

Reviewer 2: ```suggestion
using FileIO, JLD2
```

```

> We can now create a simple data frame and save it as a jld2 file using `save`. `save` 

Reviewer 2: ```suggestion
We can now create a simple data frame and save it as a JLD2 file using `save`. `save` 
```

> accepts an AbstractDict yielding the key/value pairs, where the key is a string representing 

Reviewer 2: ```suggestion
accepts an `AbstractDict` yielding the key/value pairs, where the key is a string representing 
```

the name of the dataset and the value represents its contents:

```julia
df = DataFrame(x=1, y=2)
save("output.jld2", Dict("df" => df))
```

> A jld2 file can be read in using `load`. If `load` is called with a single dataset name, 
> load returns the contents of that dataset from the file:

Reviewer 2: ```suggestion
A JLD2 file can be read in using `load`. If `load` is called with a single dataset name, 
it returns the contents of that dataset from the file:
```

```julia

> df = load("output.jld2", "df")

Reviewer 2: ```suggestion
julia> df = load("output.jld2", "df")
```

1×2 DataFrame
 Row │ x      y     
     │ Int64  Int64 
─────┼──────────────
   1 │     1      2
```

> If `load` is called with a filename argument only, the load function loads all datasets 
> from the given file into a Dict:

Reviewer 2: ```suggestion
If `load` is called with a filename argument only, it loads all objects 
from the given file into a `Dict`:
```

```julia

> df = load("output.jld2")
> Dict{String, Any} with 1 entry:
>   "df" => 1×2 DataFrame…

Reviewer 2: ```suggestion
julia> dict = load("output.jld2")
Dict{String, Any} with 1 entry:
  "df" => 1×2 DataFrame…

julia> df = dict.df
1×2 DataFrame
 Row │ x      y     
     │ Int64  Int64 
─────┼──────────────
   1 │     1      2
```
